### PR TITLE
Adds react-native-windows UWP support

### PR DESCRIPTION
**Pull request by rozele**
This adds basic support for capturing photos and videos from either the front or back panel cameras, with some support for video quality, orientation, etc. This also uses ZXing.Net to support barcode scanning from the video preview frames. It also supports torch and flash modes. Videos and photos saved to disk (or camera roll / temporary folder) also supports some file metadata (e.g., lat/long).

There are a number of features that have not yet been implemented:
- Support `playSoundOnCapture` for default shutter sounds
- Add orientation metadata properties on photo / video files
- Support all barcode formats in ZXing.Net
- Additional file metadata (like description)
- Photo quality settings with `quality` and `jpegQuality`
- Image post-processing with `mirrorImage` and `fixOrientation`
- Support event listeners for `onZoomChanged` & `onFocusChanged`
- Device authorization checks as supported on iOS

### CODE REVIEW

File: Example/package.json [hunk 1]
```diff
@@ -6,8 +6,12 @@
     "start": "node node_modules/react-native/local-cli/cli.js start"
   },
   "dependencies": {
-    "react": "~15.4.0",
+    "react": "^15.4.0",
     "react-native": "~0.40.0",
-    "react-native-camera": "file:../"
+    "react-native-camera": "~0.10.1",
+    "react-native-windows": "^0.40.0"
```
Comment (rozele) [inline Example/package.json]: Reset.

File: package.json [hunk 2]
```diff
@@ -20,7 +20,8 @@
     "barcode"
   ],
   "dependencies": {
-    "prop-types": "^15.5.10"
+    "prop-types": "^15.5.10",
+    "react": "^15.4.0"
   },
```
Comment (rozele) [inline package.json]: Delete

File: package.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
     "type": "git",
     "url": "https://github.com/lwansbrough/react-native-camera.git"
   },
-  "version": "0.10.0",
+  "version": "0.10.1",
   "description": "A Camera component for React Native. Also reads barcodes.",
```
Comment (rozele) [inline package.json]: Reset

File: windows/RNCamera/RCTCameraModule.cs [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+            using (token.Register(cancellationTokenSource.Cancel))
+            using (cancellationTokenSource.Token.Register(async () => await StopRecordingAsync(mediaCapture, taskCompletionSource)))
+            {
+                var quality = (VideoEncodingQuality)options.Value<int>("quality");
+                var encodingProfile = MediaEncodingProfile.CreateMp4(quality);
+
+                mediaCapture.AudioDeviceController.Muted = options.Value<bool>("audio");
+
+                // TODO: test orientation
+                var orientation = options.ContainsKey("orientation") ? options.Value<int>("orientation") : cameraForView.Orientation;
```
Comment (rozele) [inline windows/RNCamera/RCTCameraModule.cs]: Delete.

File: Example/package.json [hunk 5]
```diff
@@ -6,8 +6,12 @@
     "start": "node node_modules/react-native/local-cli/cli.js start"
   },
   "dependencies": {
-    "react": "~15.4.0",
+    "react": "^15.4.0",
```
Comment (rozele) [inline Example/package.json]: >^ [](start = 14, length = 1)

Reset.

**Comment by ptomasroos** [comment]
Just great @rozele ! Since I'm running vmware and windows from time to time I will def make sure to help out reviewing this!

**Comment by rt2zz** [comment]
awesome! with the number of platform support inconsistencies across iOS, android and windows, I am thinking we are going to need a docs overhaul to better communicate what works where.

**Comment by sm2017** [comment]
@ptomasroos When you accept the PR?

**Comment by sibelius** [comment]
@rozele as I don't use windows much, I've added you as a contributor

so we can help us maintain this code, feel free to merge this if it is stable enough

File: Example/package.json (hunk 5 above)
Comment (DavidRajPaul) [reply Example/package.json]: Hi...Can you please share the steps to integrate this library in react-native-windows?

**Comment by DavidRajPaul** [comment]
Hi...Can you please share the steps to integrate this library in react-native-windows?

**Comment by DavidRajPaul** [comment]
Can i Just access RNCamera from react-native-camera for windows?

**Comment by DavidRajPaul** [comment]
I followed this code to implement barcode support in react-native-windows. But got this error when importing Camera. @rozele Can you please help?
capture

**Comment by DavidRajPaul** [comment]
Any updates??

**Comment by sibelius** [comment]
RNCamera is only for RN for now, u need to use RCTCamera on Windows

I don't like how to link this on windows

**Comment by DavidRajPaul** [comment]
Thanks for the update. Can anyone help me with linking with windows or suggest a plugin to implement barcode scan in windows? Any tutorial link or proper documentation? because the implementation steps mentioned are not clear and also very old.

**Comment by kozorezal** [comment]
Seems like this version won't work with react-native-windows above 0.52 because of `System.Reactive.Linq` dependency which has version 3.1.1, but react-native-windows updated to 4.0.0 preview in 0.53.

**Comment by sibelius** [comment]
@kozorezal can u try to update this branch to fix this? or send a new pr ?

File: Example/package.json (hunk 5 above)
Comment (hoangtuandinh77) [reply Example/package.json]: https://shift.infinite.red/getting-started-with-react-native-development-for-windows-ba23a4be90b6
 check it out!

File: windows/RNCamera/RCTCameraModule.cs (hunk 4 above)
Comment (hoangtuandinh77) [reply windows/RNCamera/RCTCameraModule.cs]: image
this properties is null when i'm try to call the camera!! Can u help me??

**Comment by surajGoverdhanam** [comment]
Can anyone please provide documentation for react-native-camera windows support

**Comment by surajGoverdhanam** [comment]
I am getting "unable to get property 'Aspect ' of undefined or null".
I am using 

    "react-native": "0.54.0",
    "react-native-camera": "^1.0.2",

**Comment by gtRfnkN** [comment]
@rozele, thank you very much for this effort - the library works really well on Windows and reads QR codes just fine! I'm running `react@16.3.1` & `react-native@0.55.4`.

I am not sure, **what is actually blocking this from being merged** - can someone elaborate on that? I would love to help out with this PR, even though I have no experience with Windows programming, but I will try my best.

**For everyone being stuck with the `Aspect` property bug**: this should not happen if you're on the correct repository & branch and have set-up the library for windows correctly. Here is how you can do that:
1. add the following dependency into you `package.json`: 
`"react-native-camera": "https://github.com/rozele/react-native-camera/#windows",`.
This is the branch from @rozele with all the stuff RNWindows needs
2. link the library as described here: 
https://github.com/Microsoft/react-native-windows/blob/master/docs/LinkingLibrariesWindows.md
In the `MainReactNativeHost.cs` add `using RNCamera;` and `new RNCameraPackage()`
3. add the camera & microphone capabilities to use the camera as described here:
https://docs.microsoft.com/en-us/windows/uwp/audio-video-camera/simple-camera-preview-access#add-capability-declarations-to-the-app-manifest
4. import and used as described in `react-native-camera`

**Comment by sibelius** [comment]
Can we merge?

**Comment by gtRfnkN** [comment]
I would say not yet, I was too optimistic after it actually showed the camera stream and returned data for a QR code.

But the PR is still missing some things - here are two that I noticed already:
- add to README how to set the library up with windows (a bit like what I wrote up there)
- `onBarCodeRead` returns the data in a prop called `text` instead of `data`

I will try this library out some more and will also try to contribute as much as I can so we can get Windows supported 🙂

**Comment by expectpin** [comment]
@gtRfnk, on UWP app when I call takePictureAsync I get error '[JS Info] 'err: ', [TypeError: Object doesn't support property or method 'takePictureAsync']'.

`import Camera from 'react-native-camera';

takePicture = async () => {
    try {
      const data = await this.camera.takePictureAsync();
      console.log('Path to image: ' + data.uri);
    } catch (err) {
      console.log('err: ', err);
    }
  };

  render() {
    return (
      <View style={styles.container}>             
        <Camera
            ref={ref => {
              this.camera = ref;
            }}
            style = {styles.preview}
            permissionDialogTitle={'Permission to use camera'}
            permissionDialogMessage={'We need your permission to use your camera phone'}
        />
        <View style={{flex: 0, flexDirection: 'row', justifyContent: 'center',}}>
        <TouchableOpacity
            onPress={this.takePicture}
            style = {styles.capture}
        >
            <Text style={{fontSize: 14}}> SNAP </Text>
        </TouchableOpacity>
        </View>
      </View>
    );
  }
}`

What I'm doing wrong?

**Comment by sibelius** [comment]
Open issues if something in this PR is not working well

We can improve it in another PRs

**Comment by npmun** [comment]
@gtRfnkN @expectpin any thoughts on what I am doing wrong here? I can't make react-native-camera work on UWP for some reason.
image
`import React, { Component } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import Camera from 'react-native-camera';

export class CameraComp extends Component {
  _takePicture = async () => {
    if (this.camera) {
      const options = { quality: 0.5, base64: true };
      const data = await this.camera.takePictureAsync(options);
      console.log(data.uri);
    }
  };

  render() {
    return (
      <View style={styles.container}>
        <Camera
          ref={ref => {
            this.camera = ref;
          }}
          permissionDialogTitle={'Permission to use camera'}
          permissionDialogMessage={'We need your permission to use your camera phone'}
          style={styles.preview}
        >
          <View style={{ flex: 0, flexDirection: 'row', justifyContent: 'center' }}>
            <Text style={styles.capture} onPress={() => this._takePicture()}>
              [CAPTURE]
            </Text>
          </View>
        </Camera>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    backgroundColor: 'black'
  },
  preview: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center'
  },
  capture: {
    flex: 0,
    backgroundColor: '#fff',
    borderRadius: 5,
    padding: 15,
    paddingHorizontal: 20,
    alignSelf: 'center',
    margin: 20
  }
});`

thanks for your help!

**Comment by npmun** [comment]
nevermind...looks like Camera cannot have any children.

**Comment by vikmis007** [comment]
@rozele Is there any further update in this? I want to scan a barcode that needs to be integrated into a UWP app.

**Comment by vikmis007** [comment]
@npmun Is the camera access working for you for UWP app?

**Comment by npmun** [comment]
@vikmis007 Yes it does. But my use case is very basic.

**Comment by sibelius** [comment]
as RCTCamera was removed, you can send a PR renaming RCTCamera usage on UWP to RNCamera and it will probably work

**Pull request merged**